### DISCUSSION TITLE
Intermediate type tracing and analysis

### DISCUSSION THREAD

**Discussion by bdemann** [microsoft/pyright#5438] [Q&A]
I have a project that requires doing analysis of Python and TypeScript code.

For example if I have the following code

```python
from typing import TypedDict

BaseType = TypedDict
OtherBaseType = TypedDict
Thing1 = BaseType
Thing2 = Thing1
Thing3 = Thing2
Thing4 = TypedDict
Thing5 = Thing4
Thing6 = OtherBaseType
```

I want to be able to compare `Thing1`, `Thing2`, `Thing3`, `Thing4`, `Thing5`, and `Thing6`. Using built in Python modules I've been able to programmatically figure out that all of them are aliases of `TypedDict`, but I don't want to trace the type back that far. I want to be able to distinguish between those things that are aliases of `BaseType` (`Thing1`, `Thing2`, and `Thing3`), those that are aliases to `OtherBaseType` (`Thing6`) and those that are not either (`Thing4` and `Thing5`), and I need to be able to figure that out even if these assignments are scattered across multiple modules, and/or undergo renaming as they are imported.

So given the above source code I would like a function that would return a list of all alias to `BaseType` and a list of all aliases to `OtherBaseType`.

With TypeScript it was straight forward (see notes below for a more detailed explanation): I would get a symbol that represents `Thing3` and it would say essentially "`Thing3` refers to `Thing2`", I could then get the symbol for `Thing2` and keep working recursively until I found that `Thing1` refers to `BaseType` and then I was done. I could return `true` and move onto the next symbol.

I have not been able to find any built in Python modules that offer that kind of functionality, but it seems like Pylance/pyright ought to be able to do something like this, but I haven't had any luck figuring it out.

Is this something I could get the pyright cli to do? Or could someone point me to a part of the project that I could reverse engineer so I could make an API for it?

While I would prefer to find a library written in JavaScript, Rust, or Python, I'm happy to work in any language that has the tools I need to get this done, if someone can recommend something to me.
___
Notes:
In case it is helpful in illustrating what I am after, here are more specific details of the TypeScript tools:

I was able to import the `typescript` module from npm and call `getTypeChecker()` on a `typescript.Program` and then I was able to call  `getSymbolAtLocation()` on the resulting `typescript.TypeChecker` and pass in a `typescript.SourceFile`

Among many other things the symbol for the `typepscript.SourceFile` included a `typescript.SymbolTable` which had all of the information about all of the symbols in that source file.

I was able to inspect those symbols and if it was an imported symbol, I could follow it to the file that exported it, pull up that second files symbol table, figure out what that symbol was in the second file, and if it was an imported symbol I could keep following it back until I got to its base class.

Here is a simplified version of the symbols I got from TypeScript
```typescript
{
    name: Thing1,
    declarations: [
        {
            kind: typescript.TypeAliasDeclaration,
            type: {
                kind: typscript.Identifier,
                typeName: 'Thing2'
            }
        }
    ]
}
```
The `typescript.SymbolTable` is a dictionary with the keys `"Thing1"`, `"Thing2"`, etc and the values would be the corresponding symbol as shown above. So I could do something like:
```typescript 
const subType = symbolTable.get("Thing1").declarations[0].type.typeName;
const subTypeSymbol = symbolTable.get(subType);
```
That would give me the symbol for `Thing2` which I could then do some further analysis of to determine that it was an alias for `Thing3` etc.

**Comment by erictraut** [comment]
TypeScript offers a supported API that you can call for this sort of thing. Pyright does not provide any such API. Of course, you can fork pyright and build your own tool with the source, but there's no guarantee that the internal interfaces will remain stable over time. In fact, it's likely that they will change. 

The information you're looking for is available in pyright, but you would need to dig into the source code to find it. The `HoverProvider.ts` module might be a good place to start. I'm willing to give you some basic pointers, but since this isn't a use case for which pyright is designed, you're going to need to figure most of it out on your own.

If you prefer to work with Python code, you might want to look at mypy, which is another open-source Python type analyzer.

**Comment by erictraut** [comment]
TypeScript offers a supported API that you can call for this sort of thing. Pyright does not provide any such API. Of course, you can fork pyright and build your own tool with the source, but there's no guarantee that the internal interfaces will remain stable over time. In fact, it's likely that they will change. 

The information you're looking for is available in pyright, but you would need to dig into the source code to find it. The `HoverProvider.ts` module might be a good place to start. I'm willing to give you some basic pointers, but since this isn't a use case for which pyright is designed, you're going to need to figure most of it out on your own.

If you prefer to work with Python code, you might want to look at mypy, which is another open-source Python type analyzer.

**Comment by erictraut** [comment]
TypeScript offers a supported API that you can call for this sort of thing. Pyright does not provide any such API. Of course, you can fork pyright and build your own tool with the source, but there's no guarantee that the internal interfaces will remain stable over time. In fact, it's likely that they will change. 

The information you're looking for is available in pyright, but you would need to dig into the source code to find it. The `HoverProvider.ts` module might be a good place to start. I'm willing to give you some basic pointers, but since this isn't a use case for which pyright is designed, you're going to need to figure most of it out on your own.

If you prefer to work with Python code, you might want to look at mypy, which is another open-source Python type analyzer.
